File: emu-radio/common-library/query.cc
/*
 * Copyright (c) 2017 Cisco and/or its affiliates.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "query.h"

const std::string QueryKeys::ACTION = "action";
const std::string QueryKeys::OBJECT_NAME = "object_name";
const std::string QueryKeys::FILTER = "filter";
const std::string QueryKeys::PARAMS = "params";
const std::string QueryKeys::FIELD_NAMES = "field_names";
const std::string QueryKeys::LAST = "last";

Query::Query ()
    : query (Json::object ())
{
}

Query::Query (const std::string &action, const std::string &objectName, const std::list<std::vector<std::string>> &filter, const std::map<std::string, std::string> &params, const std::list<std::string> &fields, bool last)
{
  this->action = action;
  this->objectName = objectName;
  this->filter = filter;
  this->params = params;
  this->fields = fields;
  this->last = last;

  if (fields.size () == 0)
    {
      this->fields.push_back ("*");
    }

  query[QueryKeys::ACTION] = action;
  query[QueryKeys::OBJECT_NAME] = objectName;
  query[QueryKeys::FILTER] = Json (filter);
  query[QueryKeys::PARAMS] = Json (params);
  query[QueryKeys::FIELD_NAMES] = Json (fields);
  query[QueryKeys::LAST] = last;
}

Query Query::fromJsonString (const std::string &jsonString)
{
  Query query;
  Json jsonQuery = Json::parse (jsonString);

  std::cout << jsonQuery << std::endl;

  std::cout << jsonQuery[QueryKeys::LAST] << std::endl;

  query.setAction (jsonQuery[QueryKeys::ACTION]);
  query.setLast (jsonQuery[QueryKeys::LAST]);
  query.setObjectName (jsonQuery[QueryKeys::OBJECT_NAME]);

  Json list = jsonQuery[QueryKeys::FIELD_NAMES];

  for (Json::iterator it = list.begin (); it != list.end (); ++it)
    {
      query.fields.push_back (*it);
    }

  if (query.fields.size () == 0)
    {
      query.fields.push_back ("*");
    }

  Json list2 = jsonQuery[QueryKeys::FILTER];

  for (Json::iterator it = list2.begin (); it != list2.end (); ++it)
    {
      query.filter.push_back (*it);
    }

  Json map = jsonQuery[QueryKeys::PARAMS];

  for (Json::iterator it = map.begin (); it != map.end (); ++it)
    {
      std::cout << it.key () << " " << it.value ().dump () << std::endl;
      query.params[it.key ()] = it.value ().dump ();
    }

  query.query[QueryKeys::ACTION] = query.action;
  query.query[QueryKeys::OBJECT_NAME] = query.objectName;
  query.query[QueryKeys::FILTER] = Json (query.filter);
  query.query[QueryKeys::PARAMS] = Json (query.params);
  query.query[QueryKeys::FIELD_NAMES] = Json (query.fields);
  query.query[QueryKeys::LAST] = query.last;

  return query;
}

void Query::removeQuotes (std::string &string)
{
  std::cout << string.front () << std::endl;
  string.erase (remove (string.begin (), string.end (), '\"'), string.end ());
}

std::string Query::toJsonString (const Query &query)
{
  return query.query.dump ();
}

std::string
Query::toJsonString (const std::string &action, const std::string &objectName, const std::list<std::vector<std::string>> &filter, const std::map<std::string, std::string> &params, const std::list<std::string> &fields, bool last)
{
  Json jsonQuery;

  jsonQuery[QueryKeys::ACTION] = action;
  jsonQuery[QueryKeys::OBJECT_NAME] = objectName;
  jsonQuery[QueryKeys::FILTER] = Json (filter);
  jsonQuery[QueryKeys::PARAMS] = Json (params);
  jsonQuery[QueryKeys::FIELD_NAMES] = Json (fields);
  jsonQuery[QueryKeys::LAST] = last;

  return jsonQuery.dump ();
}

std::string Query::toJsonString ()
{
  return query.dump ();
}

const std::string &Query::getAction () const
{
  return action;
}

void Query::setAction (const std::string &action)
{
  Query::action = action;
  removeQuotes (this->action);
  query[QueryKeys::ACTION] = this->action;
}

const std::string &Query::getObjectName () const
{
  return objectName;
}

void Query::setObjectName (const std::string &objectName)
{
  Query::objectName = objectName;
  removeQuotes (this->objectName);
  query[QueryKeys::OBJECT_NAME] = this->objectName;
}

const std::list<std::vector<std::string>> &Query::getFilter () const
{
  return filter;
}

void Query::setFilter (const std::list<std::vector<std::string>> &filter)
{
  Query::filter = filter;

  for (auto f : this->filter)
    {
      for (auto field : f)
        {
          removeQuotes (field);
        }
    }

  query[QueryKeys::FILTER] = this->filter;
}

const std::map<std::string, std::string> &Query::getParams () const
{
  return params;
}

void Query::setParams (const std::map<std::string, std::string> &params)
{
  Query::params = params;
  query[QueryKeys::PARAMS] = this->params;
}

const std::list<std::string> &Query::getFields () const
{
  return fields;
}

void Query::setFields (const std::list<std::string> &fields)
{
  Query::fields = fields;

  for (auto field : this->fields)
    {
      removeQuotes (field);
    }

  query[QueryKeys::FIELD_NAMES] = this->fields;
}

bool Query::isLast () const
{
  return last;
}

void Query::setLast (int last)
{
  Query::last = last;
  query[QueryKeys::LAST] = this->last;
}

bool Query::isEmpty ()
{
  return query.empty ();
}

File: netmodel/model/query.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (c) 2017 Cisco and/or its affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from netmodel.model.filter          import Filter
from netmodel.model.field_names     import FieldNames

ACTION_INSERT    = 1
ACTION_SELECT    = 2
ACTION_UPDATE    = 3
ACTION_DELETE    = 4
ACTION_EXECUTE   = 5
ACTION_SUBSCRIBE = 6
ACTION_UNSUBSCRIBE = 7

ACTION2STR = {
    ACTION_INSERT    : 'insert',
    ACTION_SELECT    : 'select',
    ACTION_UPDATE    : 'update',
    ACTION_DELETE    : 'delete',
    ACTION_EXECUTE   : 'execute',
    ACTION_SUBSCRIBE : 'subscribe',
    ACTION_UNSUBSCRIBE : 'unsubscribe',
}
STR2ACTION = dict((v, k) for k, v in ACTION2STR.items())

FUNCTION_SUM      = 1

FUNCTION2STR = {
    FUNCTION_SUM     : 'sum'
}
STR2FUNCTION = dict((v, k) for k, v in FUNCTION2STR.items())

class Query:
    def __init__(self, action, object_name, filter = None, params = None,
            field_names = None, aggregate = None, last = False, reply = False):
        self.action      = action
        self.object_name = object_name

        if filter:
            if isinstance(filter, Filter):
                self.filter = filter
            else:
                self.filter = Filter.from_list(filter)
        else:
            self.filter = Filter()

        self.params = params

        if field_names:
            if isinstance(field_names, FieldNames):
                self.field_names = field_names
            else:
                self.field_names = FieldNames(field_names)
        else:
            self.field_names = FieldNames()

        self.aggregate = aggregate

        self.last = last
        self.reply = reply

    def to_dict(self):
        aggregate = FUNCTION2STR[self.aggregate] if self.aggregate else None
        return {
            'action': ACTION2STR[self.action],
            'object_name': self.object_name,
            'filter': self.filter.to_list(),
            'params': self.params,
            'field_names': self.field_names,
            'aggregate': aggregate,
            'reply': self.reply,
            'last': self.last
        }

    @staticmethod
    def from_dict(dic):
        action = STR2ACTION[dic.get('action').lower()]
        object_name = dic.get('object_name')
        filter = dic.get('filter', None)
        params = dic.get('params', None)
        field_names = dic.get('field_names', None)
        aggregate = STR2FUNCTION[dic.get('aggregate').lower()] \
                    if dic.get('aggregate') else None
        if field_names  == '*':
            field_names = FieldNames(star = True)
        last = dic.get('last', False)
        reply = dic.get('reply', False)
        return Query(action, object_name, filter, params, field_names,
                aggregate, last)

    def to_sql(self, multiline = False):
        """
        Args:
            platform: A String corresponding to a namespace (or platform name)
            multiline: A boolean indicating whether the String could contain
                carriage return.
        Returns:
            The String representing this Query.
        """
        get_params_str = lambda : ", ".join(["%s = %r" % (k, v) \
                for k, v in self.params.items()])

        object_name  = self.object_name
        field_names = self.field_names
        field_names_str = ('*' if field_names.is_star() \
                else ', '.join([field for field in field_names]))
        select = "SELECT %s" % ((FUNCTION2STR[self.aggregate] + "(%s)") \
                if self.aggregate else '%s') % field_names_str
        filter  = "WHERE %s" % self.filter      if self.filter           else ''
        #at     = "AT %s"     % self.get_timestamp() if self.get_timestamp() else ""
        at = ''
        params = "SET %s"    % get_params_str()     if self.params       else ''

        sep = " " if not multiline else "\n  "

        strmap = {
            ACTION_SELECT : "%(select)s%(sep)s%(at)s%(sep)sFROM %(object_name)s%(sep)s%(filter)s",
            ACTION_UPDATE : "UPDATE %(object_name)s%(sep)s%(params)s%(sep)s%(filter)s%(sep)s%(select)s",
            ACTION_INSERT : "INSERT INTO %(object_name)s%(sep)s%(params)s",
            ACTION_DELETE : "DELETE FROM %(object_name)s%(sep)s%(filter)s",
            ACTION_SUBSCRIBE : "SUBSCRIBE : %(select)s%(sep)s%(at)s%(sep)sFROM %(object_name)s%(sep)s%(filter)s",
            ACTION_UNSUBSCRIBE : "UNSUBSCRIBE : %(select)s%(sep)s%(at)s%(sep)sFROM %(object_name)s%(sep)s%(filter)s",
            ACTION_EXECUTE : "EXECUTE : %(select)s%(sep)s%(at)s%(sep)sFROM %(object_name)s%(sep)s%(filter)s",
        }

        return strmap[self.action] % locals()

    def __repr__(self):
        return self.to_sql()

    __str__ = __repr__
